lighthouse-core/gather/gatherers/seo/font-size.js

```diff
@@ -44,58 +48,130 @@ function nodeInBody(node) {
+ * Computes the CSS specificity of a given selector, i.e. #id > .class > div
  *
- * @param {string} property CSS property name
- * @param {LH.Crdp.DOM.Node} node
+ * @see https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity
+ * @see https://www.smashingmagazine.com/2010/04/css-specificity-and-inheritance/
+ *
+ * @param {string} selector
+ * @return {number}
+ */
+function computeSelectorSpecificity(selector) {
+  const tokens = selector.split(' ');
+
+  let numIDs = 0;
+  let numClasses = 0;
+  let numTypes = 0;
+
+  for (const token of tokens) {
+    const ids = token.match(/\b#[a-z0-9]+/g) || [];
+    const classes = token.match(/\b\.[a-z0-9]+/g) || [];
+    const types = token.match(/^[a-z]+/) ? [1] : [];
+    numIDs += ids.length;
+    numClasses += classes.length;
+    numTypes += types.length;
+  }
+
+  return Math.min(9, numIDs) * 100 + Math.min(9, numClasses) * 10 + Math.min(9, numTypes);
+}
+
+/**
+ * Finds the most specific directly matched CSS rule from the list.
+ *
+ * @param {Array<LH.Crdp.CSS.RuleMatch>} [matchedCSSRules]
+ * @returns {NodeFontData['cssRule']|undefined}
+ */
+function findMostSpecificMatchedCSSRule(matchedCSSRules = []) {
+  let maxSpecificity = -Infinity;
+  /** @type {LH.Crdp.CSS.CSSRule|undefined} */
+  let maxSpecificityRule;
+
+  for (const {rule, matchingSelectors} of matchedCSSRules) {
+    if (hasFontSizeDeclaration(rule.style)) {
+      const specificities = matchingSelectors.map(idx =>
+        computeSelectorSpecificity(rule.selectorList.selectors[idx].text)
+      );
+      const specificity = Math.max(...specificities);
+      // Use greater OR EQUAL so that the last rule wins in the event of a tie
+      if (specificity >= maxSpecificity) {
+        maxSpecificity = specificity;
+        maxSpecificityRule = rule;
+      }
+    }
+  }
+
+  if (maxSpecificityRule) {
+    return {
+      type: 'Regular',
+      ...maxSpecificityRule.style,
+      parentRule: {
+        origin: maxSpecificityRule.origin,
+        selectors: maxSpecificityRule.selectorList.selectors,
+      },
+    };
+  }
+}
+
+/**
+ * Finds the most specific directly matched CSS rule from the list.
+ *
+ * @param {Array<LH.Crdp.CSS.InheritedStyleEntry>} [inheritedEntries]
+ * @returns {NodeFontData['cssRule']|undefined}
+ */
+function findInheritedCSSRule(inheritedEntries = []) {
+  // The inherited array contains the array of matched rules for all parents in ascending tree order.
+  // i.e. for an element whose path is `html > body > #main > #nav > p`
+  // `inherited` will be an array of styles like `[#nav, #main, body, html]`
+  // The closest parent with font-size will win
+  for (const {inlineStyle, matchedCSSRules} of inheritedEntries) {
+    if (hasFontSizeDeclaration(inlineStyle)) return {type: 'Inline', ...inlineStyle};
+
+    const directRule = findMostSpecificMatchedCSSRule(matchedCSSRules);
+    if (directRule) return directRule;
+  }
+}
+
+/**
+ * Returns effective CSS rule for given CSS property.
```

Comment:
Since this is just matching against `font-size`, maybe mention that instead. "effective" also isn't immediately clear here, instead of something about how it's the css rule currently setting the style for the `font-size` property

Reply:
done